### ISSUE TITLE
Collaborative p2p training

### ISSUE THREAD

**Issue by DrBellubins** [jrz97619761/test-model-thing#8]
Since latent-space and recurrent models benefit from constant recursive training, it doesn't seem like a stretch to enable distributed training amongst anyone who has a computer.

This is a proof-of-concept implementation of a peer-to-peer distributed training.

Full disclosure, this was vibe-coded, and I can't test it on my machine because I don't have an apple device.

## Summary of the Flow
1. Each node independently trains locally (byte-LM or CoLA head).
2. Periodically, a node diffs its trainable parameters against last-known hashes, packages changes into a hash-chained `UpdateBatch`, and applies it locally.
3. The batch is broadcast to connected peers over WebSocket.
4. Peers validate the batch (architecture, hash chain, sequence, tensor hashes, resulting root) before applying it to their own live model.
5. If a peer detects it has fallen behind (sequence mismatch) or received an invalid/out-of-order batch, it performs an HTTP-based chunked "repair" — pulling a verified full snapshot from the peer and resetting its local chain to match, avoiding complex merge logic entirely (last-writer/most-advanced-wins model, not CRDT-style merging).

This effectively creates a simple gossip-based, hash-verified, linear-history parameter synchronization protocol layered on top of ad hoc peer training, prioritizing integrity (via SHA-256 + root hashing at every step) over conflict resolution sophistication (forks are resolved via full snapshot repair, not merging).

Also as a bonus, it has a web-ui that shows training stats, and p2p sync stats:

```
python serve.py \
  --host 0.0.0.0 --port 8080 \
  --checkpoint smaller-4.5m.safetensors \
  --dim 512 --layers 16 --temp 0.75 --lr 5e-4 \
  --run-train-loop \
  --train-text-path path/to/corpus.txt
```

I'd love to know if this has any actual merit, just an idea I've had for awhile.

The ultimate hope is that there could be a model that gets progressively smarter as people use it, presenting it with more data, and those improvements are shared across a distributed network. This would break the train first, then inference cycle that makes the barrier to entry for training huge. This architecture currently already blurs the line between training and inference, the idea is to just network users together.

The repo: https://github.com/DrBellubins/test-model-thing

**Comment by jrz97619761** [comment]
Cool! If you want you can check out the windows forks that others have made (#1) and see if you can run those. Both aren't 100% accurate replicas but you can try.

Your idea is something I've been thinking about but haven't got around to doing. This is quite nice, thank you!

Will try it out later as well

**Comment by DrBellubins** [comment]
I want to try to port this to the pytorch forks so I can test it myself when I get the time.

This is from GPT:

# Running the Web UI + Training via `serve.py`

The `serve.py` script is the entry point — it starts a FastAPI web server (dashboard UI + sync API + WebSocket peer sync) and can optionally run background training loops. Here's how to use it.

## Basic Command

```bash
python serve.py --host 127.0.0.1 --port 8080 --checkpoint smaller-4.5m.safetensors
```

This alone starts the web UI at `http://127.0.0.1:8080/` and enables P2P sync endpoints, but **no training runs** unless you explicitly enable a training mode with a flag.

## Option 1: Byte-level Language Model Training Loop

Trains the model character-by-character over a plain text corpus in the background.

```bash
python serve.py \
  --host 0.0.0.0 --port 8080 \
  --checkpoint smaller-4.5m.safetensors \
  --dim 512 --layers 16 --temp 0.75 --lr 5e-4 \
  --run-train-loop \
  --train-text-path path/to/corpus.txt
```

- `--run-train-loop` enables `_training_task`, which reads `--train-text-path` line-by-line, feeding byte pairs through the model (`Model.__call__`) and checkpointing every 500 steps.
- If `--train-text-path` is omitted while `--run-train-loop` is set, it logs a warning and the training task exits immediately — so you must supply the path.

## Option 2: CoLA Classification Head Training/Eval

Runs the linguistic-acceptability benchmark workflow from `benchmark.py` in the background.

```bash
python serve.py \
  --host 0.0.0.0 --port 8080 \
  --checkpoint smaller-4.5m.safetensors \
  --run-cola \
  --cola-train-path CoLA/original/raw/in_domain_train.tsv \
  --cola-eval-path CoLA/original/raw/in_domain_dev.tsv \
  --cola-epochs 3
```

- `--run-cola` enables `_cola_task`, which freezes the base model, trains a small `Classification` head via `train_cola_head`, and (if `--cola-eval-path` is given) evaluates it via `evaluate_cola_head`.
- `--cola-train-path` is **required** when `--run-cola` is set; if missing, the task logs an error and exits without training.
- `--cola-epochs` controls how many passes over the training set (default `3`).

You can technically pass both `--run-train-loop` and `--run-cola` together since they're independent asyncio tasks, though running both simultaneously means they'll contend for `model_adapter.lock` (thread-safe, but slower).

## Option 3: Multi-Peer Sync

To connect nodes together for P2P weight sync, repeat `--peer` for each peer's WebSocket URL:

```bash
# Node A
python serve.py --host 0.0.0.0 --port 8080 --checkpoint node-a.safetensors --run-train-loop --train-text-path corpus.txt

# Node B — points at Node A's sync websocket
python serve.py --host 0.0.0.0 --port 8081 --checkpoint node-b.safetensors --run-train-loop --train-text-path corpus.txt \
  --peer ws://<node-a-host>:8080/sync/ws
```

Peers can be chained/meshed by adding `--peer` multiple times. Both nodes must have **matching model architecture** (`--dim`, `--layers`) — the `architecture_id` includes these plus a manifest hash, and mismatches will cause `hello` handshakes to be rejected (logged as an "architecture mismatch" warning in the event log).

If you want to protect the `/sync/*` HTTP+WS endpoints (useful when peers connect over an untrusted network), set a shared token:

```bash
--sync-token some-shared-secret
```
Peers must include this same token (as `?token=` for WS or `x-sync-token` header for HTTP repair calls) or requests get a 401.

## Viewing the Dashboard

Once running, open `http://<host>:<port>/` in a browser. The page (`HTML_PAGE` in `serve.py`) polls `/api/status` and streams `/api/events` (Server-Sent Events) to show:
- **Node** info (ID, architecture, checkpoint path)
- **Training** metrics (epoch, step, loss, throughput, CoLA MCC/confusion if applicable)
- **Synchronization** state (tip batch, sequence, connected peers, last sync time/errors)
- **Sharing Policy** (confirms only trainable params are shared — `embedtrace`, `state.*`, `decaytrace.*`, and optimizer state stay local)
- **Activity Log** — a live tail of events (peer connects, batch accept/reject, repairs, training progress)

## Full Reference of Relevant Flags

| Flag | Purpose | Default |
|---|---|---|
| `--host` / `--port` | Bind address | `127.0.0.1` / `8080` |
| `--checkpoint` | Path to `.safetensors` checkpoint file | `smaller-4.5m.safetensors` |
| `--dim` / `--layers` / `--temp` / `--lr` | Model architecture/hyperparams | `512` / `16` / `0.75` / `5e-4` |
| `--peer` | Peer WS URL (repeatable) | none |
| `--sync-interval` | Seconds between local diff/broadcast checks | `3.0` |
| `--sync-max-payload-bytes` | Max batch size in bytes | `8 MiB` |
| `--sync-chunk-size` | Chunk size for repair tensor pulls | `256 KiB` |
| `--sync-token` | Shared auth token for `/sync/*` | none |
| `--run-train-loop` + `--train-text-path` | Enable byte-LM background training | disabled |
| `--run-cola` + `--cola-train-path` [+ `--cola-eval-path`] + `--cola-epochs` | Enable CoLA head training/eval | disabled |
| `--log-level` | Logging verbosity | `INFO` |

**Note:** `serve.py`'s `main()` requires MLX to be importable (`from main import Model`) — if MLX isn't installed or isn't supported on your platform, it raises a `SystemExit` telling you to run it on a machine with a working MLX install (MLX is Apple Silicon-only in most builds).

**Comment by jrz97619761** [comment]
Nice! This is very cool. I'll check out the repo in a second. If you want you could spin up GPT to help w/ #2 as well since this is the main open issue with the code right now. Thanks!

**Comment by ceeray** [comment]
Guys, for the long term please consider training on websites such as Grokipedia as complementary or alternative sources to Wikipedia, to reduce the systemic biases inherent within the Wikipedia editorial culture and to expose the model to differently curated knowledge bases.

**Comment by reyvvon** [comment]
I don't even need to explain why this is a bad idea once you take a look at Grokipedia and find the amount of bias(in the exact opposite way of the original article instead of a middle ground) and hallucinations.

**Comment by ceeray** [comment]
Fair enough but I still think it would be good to have other source material to complement wikipedia if that's what is being used as a base to train upon.

**Comment by jrz97619761** [comment]
Never heard of "Grokipedia". Wikipedia has enough specialized topics e.g. articles on operating system development. I don't know if "Grokipedia" has the same.

Honestly I don't care for politics myself so I'd leave that to you as the user, whether to train your data on certain sources or not.

**Comment by ceeray** [comment]
Here's some more reputable alternatives fwiw.

https://www.reputationx.com/blog/wikipedia-alternatives

**Comment by jrz97619761** [comment]
Also, @ceeray, Wikipedia is just an example. At some point I'm thinking to implement RL for the model (once the memory issues are ironed out) rather than rely so much on datasets.

I'm thinking we just give the model a computer environment (perhaps in a VM) and let it experiment however. If we have some sort of system that allows the model to, say, think internally before taking an action on the screen (which would either be a mouse or keyboard input) and then reading the current screen image as bytes again. Imagine that happening at 60Hz or something and the model responding to changes in the desktop in real time, rather than having to spend minutes thinking in language like an LLM would before taking a set of programmed skill actions.

So, that would look like:

Encoder would encode the current screen, screenshotted, as a series of bytes
Decoder would decode the bytes to different inputs on a keyboard and mouse

(Perhaps we could have a mechanism for the same model to have multiple encoders and decoders directly feeding into the central trace layers at once, so the model can see the visual and hear the audio at once, or decode to keyboard and mouse at once without waiting turns).

Obviously this is an example of a longer term idea but I much prefer RL over this current training system in the current code. So, don't worry about Wikipedia, that's just a temporary thing to prove the architecture works. I think multimodality and continuous streaming is the thing to test next.

(If I made some terrible mistake when describing the idea feel free to correct me.)

**Comment by jrz97619761** [comment]
I just checked, Grokipedia is just output of Grok, if we trained on that we'd be distilling Grok. that's a bad idea because if we did, xAI would come after us.
